R/gm-auth.R

```diff
@@ -0,0 +1,209 @@
+    cache = cache,
+    use_oob = use_oob,
+    token = token
+  )
+  if (!inherits(cred, "Token2.0")) {
+    stop(
+      "Can't get Google credentials.\n",
+      "Are you running gmailr in a non-interactive session? Consider:\n",
+      "  * Call `gm_auth()` directly with all necessary specifics.\n",
+      call. = FALSE
+    )
+  }
+  .auth$set_cred(cred)
+  .auth$set_auth_active(TRUE)
+
+  invisible()
+}
+
+gm_default_email <- function() {
+  user <- Sys.getenv("GMAILR_EMAIL")
+  if (nzchar(user)) {
+    return(user)
+  }
+  NULL
+}
+
+#' Clear current token
+#'
+#' @eval gargle:::PREFIX_deauth_description_no_api_key(gargle_lookup_table)
+#'
+#' @family auth functions
+#' @export
+#' @examples
+#' \dontrun{
+#' gm_deauth()
+#' }
+gm_deauth <- function() {
+  .auth$clear_cred()
+  invisible()
+}
+
+#' Produce configured token
+#'
+#' @eval gargle:::PREFIX_token_description(gargle_lookup_table)
+#' @eval gargle:::PREFIX_token_return()
+#'
+#' @family low-level API functions
+#' @export
+#' @examples
+#' \dontrun{
+#' gm_token()
+#' }
+gm_token <- function() {
+  if (!gm_has_token()) {
+    gm_auth()
+  }
+  httr::config(token = .auth$cred)
+}
+
+#' Is there a token on hand?
+#'
+#' Reports whether gmailr has stored a token, ready for use in downstream
+#' requests.
+#'
+#' @return Logical.
+#' @export
+#'
+#' @examples
+#' gm_has_token()
+gm_has_token <- function() {
+  inherits(.auth$cred, "Token2.0")
+}
+
+#' Edit auth configuration
+#'
+#' @eval gargle:::PREFIX_auth_configure_description(gargle_lookup_table, .has_api_key = FALSE)
+#' @eval gargle:::PREFIX_auth_configure_params(.has_api_key = FALSE)
+#' @eval gargle:::PREFIX_auth_configure_return(gargle_lookup_table, .has_api_key = FALSE)
+#'
+#' @inheritParams httr::oauth_app
+#' @param secret consumer secret, also sometimes called the client secret.
+#' @param ... Additional arguments passed to \code{\link[httr:oauth_app]{httr::oauth_app()}}
+#' @family auth functions
+#' @export
+#' @examples
+#' \dontrun{
+#' # see the current user-configured OAuth app (probaby `NULL`)
+#' gm_oauth_app()
+#'
+#' if (require(httr)) {
+#'
+#'   # store current state, so we can restore
+#'   original_app <- gm_oauth_app()
+#'
+#'   # bring your own app via client id (aka key) and secret
+#'   google_app <- httr::oauth_app(
+#'     "my-awesome-google-api-wrapping-package",
+#'     key = "123456789.apps.googleusercontent.com",
+#'     secret = "abcdefghijklmnopqrstuvwxyz"
+#'   )
+#'   gm_auth_configure(app = google_app)
+#'
+#'   # confirm current app
+#'   gm_oauth_app()
+#'
+#'   # restore original state
+#'   gm_auth_configure(app = original_app)
+#'   gm_oauth_app()
+#' }
+#'
+#' # bring your own app via JSON downloaded from Google Developers Console
+#' gm_auth_configure(
+#'   path = "/path/to/the/JSON/you/downloaded/from/google/dev/console.json"
+#' )
+#' }
+gm_auth_configure <- function(key = Sys.getenv("GMAILR_APP_KEY"),
+                              secret = Sys.getenv("GMAILR_APP_SECRET"),
```

Comment:
I'm pretty interested in this, because gargle probably should have an "official" method of providing the app when the app is not optional.